### PR TITLE
[ENH] Adding score metrics

### PR DESCRIPTION
**Pull request by emdupre**
Initial commit of scoring metrics, based on discussion in #37.

To do:
- [x] appropriately handle `sample_weight` in normalized reconstruction
- [x] update documentation accordingly

### CODE REVIEW

**Comment by emdupre** [comment]
OK, I think this is ready for review !

Right now I'm still only showcasing correlations in the documentation, but I can update those to use R2 and/or reconstruction error, if you'd like ! Just wasn't sure if it should be in this PR or another one :smile_cat:

File: fmralign/metrics.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
+import numpy as np
+from scipy.stats import pearsonr
+from sklearn.metrics import r2_score
+
+
+def score_table(ground_truth, prediction, masker, loss,
+                multioutput='raw_values'):
```
Comment (bthirion) [inline fmralign/metrics.py]: 'multi_output' is more pythonic isn't it ?

File: fmralign/metrics.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            The R2 distance between source and target arrays.
+            Best possible score is 1.0 and it can be negative (because the
+            model can be arbitrarily worse).
+        'corr' :
+            The correlation between source and target arrays.
+        'n_reconstruction_err' :
+            The normalized reconstruction error. A perfect prediction
+            yields a value of 1.0
+    multioutput: str in ['raw_values', 'uniform_average']
+        Defines aggregating of multiple output scores. Default is raw values.
```
Comment (bthirion) [inline fmralign/metrics.py]: applies only if loss is not 'corr'

File: fmralign/metrics.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    Parameters
+    ----------
+    y_true : arr
+        The ground truth array.
+    y_pred : arr
+        The predicted array.
+    sample_weights : arr
+        Weights to assign to each sample.
+    multioutput: str in ['raw_values', 'uniform_average',
+                         'variance_weighted']
```
Comment (bthirion) [inline fmralign/metrics.py]: is variance_weighted used anywhere ?

File: fmralign/metrics.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                                      denominator[valid_score])
+    if multioutput == 'raw_values':
+        # return scores individually
+        return output_scores
+    
+    elif multioutput == 'uniform_average':
+        # passing None as weights yields uniform average
+        return np.average(output_scores, weights=None)
+
+    elif multioutput == 'variance_weighted':
```
Comment (bthirion) [inline fmralign/metrics.py]: Unclear to me what the use case for that is.

File: fmralign/metrics.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    `Bazeille T., Richard H., Janati H., and Thirion B. (2019) Local
+    Optimal Transport for Functional Brain Template Estimation.
+    In: Chung A., Gee J., Yushkevich P., and Bao S. (eds) Information 
+    Processing in Medical Imaging. Lecture Notes in Computer Science,
+    vol 11492. Springer, Cham.
+    DOI: 10.1007/978-3-030-20351-1_18.`
+    """
+    num = 1 - aligned_error
+    den = 1 - identity_error
+    return 1 - (num / den)
```
Comment (bthirion) [inline fmralign/metrics.py]: no need to handle den=0 ?

File: fmralign/metrics.py (hunk 1 above)
Comment (thomasbazeille) [reply fmralign/metrics.py]: multioutput for consistency with api of sklearn r2_score (and metrics in general)

File: fmralign/metrics.py (hunk 2 above)
Comment (thomasbazeille) [reply fmralign/metrics.py]: no apply for all metrics since they are all voxelwise (with time dimension as samples) and the averaging here is across voxels)

File: fmralign/metrics.py (hunk 3 above)
Comment (thomasbazeille) [reply fmralign/metrics.py]: no i'm not sure if it'd be useful but this option makes the API fully compatible with sklearn's r2_score

File: fmralign/metrics.py (hunk 4 above)
Comment (thomasbazeille) [reply fmralign/metrics.py]: see above

**Comment by thomasbazeille** [comment]
I don't know why CircleCI is not running, I'll look into this

File: fmralign/metrics.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
+import numpy as np
+from scipy.stats import pearsonr
+from sklearn.metrics import r2_score
+
+
+def score_table(ground_truth, prediction, masker, loss,
```
[the remark this answers is not in the document]
Comment (emdupre) [reply fmralign/metrics.py]: I went with `score_voxelwise` -- let me know what you think !

File: fmralign/metrics.py (hunk 5 above)
Comment (emdupre) [reply fmralign/metrics.py]: Should be handled now :smile_cat: Right now I'm returning 0 if den is 0, but let me know if there's another solution you'd prefer !

**Comment by emdupre** [comment]
Thanks both for your helpful feedback, here ! I think I've integrated most of it. I also kept accidentally adding the wrong files, so I went ahead and added a `.gitignore` to the project. If you'd like that in another PR let me know and I can cherrypick the commit !

I don't have strong opinions on `variance_weighted` -- I agree that it's nice to align with sklearn's interface, but I'm struggling to see a use case for it. I can try it out in the documentation if you'd like ?

File: fmralign/metrics.py (hunk 3 above)
Comment (bthirion) [reply fmralign/metrics.py]: OK, maybe plainly state it that way.

File: fmralign/metrics.py (hunk 2 above)
Comment (bthirion) [reply fmralign/metrics.py]: But if I'm not mistaken this is not used in the code (when loss == 'corr'), is it ?

File: fmralign/metrics.py (hunk 4 above)
Comment (bthirion) [reply fmralign/metrics.py]: OK. It's always a bad  to have some unused / ill-motivated features, but i have no better idea.

File: fmralign/metrics.py (hunk 2 above)
Comment (thomasbazeille) [reply fmralign/metrics.py]: Yes you're right, only the multioutput case is currently handled

File: fmralign/metrics.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    X_pred = masker.transform(prediction)
+
+    if loss is "R2":
+        score = r2_score(X_gt, X_pred, multioutput=multioutput)
+    elif loss is "n_reconstruction_err":
+        score = normalized_reconstruction_error(
+            X_gt, X_pred, multioutput=multioutput)
+    elif loss is "corr":
+        score = np.array([pearsonr(X_gt[:, vox], X_pred[:, vox])[0]  # pearsonr returns both rho and p
+                          for vox in range(X_pred.shape[1])])
```
Comment (thomasbazeille) [inline fmralign/metrics.py]: Related to B.T comment above, multioutput not used

```
if multioutput == "uniform_average":
            score = np.mean(score)
```

should do the job, even though I didn't tested it. It's a pity tests didn't get that, we're not testing uniform_average right ?

Comment (emdupre) [reply fmralign/metrics.py]: It doesn't look like CI tests are running at all, currently. Should they be ? pytest is passing locally but I'll add a shape check so the previous commit would have failed !

**Comment by thomasbazeille** [comment]
Sorry for the delay I updated our CircleCI config, pushing an empty commit should run the tests!

**Comment by emdupre** [comment]
> Sorry for the delay I updated our CircleCI config, pushing an empty commit should run the tests!

I pushed something but I don't see anything yet :disappointed: 

Let me know if I can help with the circle config !!

**Comment by thomasbazeille** [comment]
Sorry this time it could work.

**Comment by emdupre** [comment]
Yayyy tests ! :tada: :sparkles: 

I think this is ready for a final review :smile_cat:

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,127 @@
+# Byte-compiled / optimized / DLL files
```
Comment (bthirion) [inline .gitignore]: Not a big fan of that (don't have it in my other projects), but no strong feelings.

File: fmralign/metrics.py (hunk 2 above)
Comment (bthirion) [reply fmralign/metrics.py]: I think that the docstring should be a bit mroe rpecise here.

Comment (thomasbazeille) [reply fmralign/metrics.py]: it was added in the code since then so now everything is consistent and docstring is right I think

File: fmralign/metrics.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    Parameters
+    ----------
+    y_true : arr
+        The ground truth array.
+    y_pred : arr
+        The predicted array.
+    sample_weights : arr
+        Weights to assign to each sample.
+    multioutput: str in ['raw_values', 'uniform_average',
+                         'variance_weighted']
```
Comment (thomasbazeille) [inline fmralign/metrics.py]:
```suggestion
]
```

File: fmralign/metrics.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    
+    Parameters
+    ----------
+    y_true : arr
+        The ground truth array.
+    y_pred : arr
+        The predicted array.
+    sample_weights : arr
+        Weights to assign to each sample.
+    multioutput: str in ['raw_values', 'uniform_average',
```
Comment (thomasbazeille) [inline fmralign/metrics.py]:
```suggestion
    multioutput: str in ['raw_values', 'uniform_average'
```

File: fmralign/metrics.py (hunk 9 above)
Comment (thomasbazeille) [inline fmralign/metrics.py]:
```suggestion
]
```

Comment (emdupre) [reply fmralign/metrics.py]: I committed this outside of the UI so resolving ! :smile_cat:

**Comment by thomasbazeille** [comment]
Any more comments @bthirion  ?
